src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PETypeParameterSymbol.cs

```diff
@@ -234,31 +254,44 @@ public override bool HasConstructorConstraint
         {
             get
             {
-                return (_flags & GenericParameterAttributes.DefaultConstructorConstraint) != 0;
+                GetDeclaredConstraintTypes();
+                return (_lazyFlags & GenericParameterAttributes.DefaultConstructorConstraint) != 0;
             }
         }
 
         public override bool HasReferenceTypeConstraint
         {
             get
             {
-                return (_flags & GenericParameterAttributes.ReferenceTypeConstraint) != 0;
+                GetDeclaredConstraintTypes();
+                return (_lazyFlags & GenericParameterAttributes.ReferenceTypeConstraint) != 0;
             }
         }
 
         public override bool HasValueTypeConstraint
         {
             get
             {
-                return (_flags & GenericParameterAttributes.NotNullableValueTypeConstraint) != 0;
+                GetDeclaredConstraintTypes();
+                return (_lazyFlags & GenericParameterAttributes.NotNullableValueTypeConstraint) != 0 || HasUnmanagedTypeConstraint;
```

Comment:
>|| HasUnmanagedTypeConstraint [](start = 101, length = 29)

Presence of the attribute should not imply any constraints, whatever is specified in metadata flags should define constraints.